pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -316,6 +316,242 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
+		}
+	}
+
+	// Check for empty shared pool for concurrently running containers (app and restartable init containers).
+	hasSharedAppContainers := false
+	for _, container := range pod.Spec.Containers {
+		if container.Resources.Requests.Cpu().IsZero() {
+			hasSharedAppContainers = true
+			break
+		}
+	}
+	if !hasSharedAppContainers {
+		for _, container := range pod.Spec.InitContainers {
+			if !podutil.IsRestartableInitContainer(&container) {
+				continue
+			}
+			if container.Resources.Requests.Cpu().IsZero() {
+				hasSharedAppContainers = true
+				break
+			}
+		}
+	}
+
+	if sumOfLongRunningExclusiveCPUs >= podTotalCPUs && hasSharedAppContainers {
+		return fmt.Errorf("pod rejected, sum of exclusive container cpu requests equals pod budget, leaving no cpus for shared containers")
+	}
+
+	// Check for empty shared pool for standard init containers.
+	sumOfSidecarExclusiveCPUs := 0
+	for _, container := range pod.Spec.InitContainers {
+		if podutil.IsRestartableInitContainer(&container) {
+			sumOfSidecarExclusiveCPUs += p.guaranteedCPUs(logger, pod, &container)
+		}
+	}
+
+	poolForStandardInits := podTotalCPUs - sumOfSidecarExclusiveCPUs
+	maxStandardInitExclusiveCPUs := 0
+	hasSharedStandardInitContainer := false
+	for _, container := range pod.Spec.InitContainers {
+		if !podutil.IsRestartableInitContainer(&container) {
+			cpus := p.guaranteedCPUs(logger, pod, &container)
+			if cpus > 0 {
+				if cpus > maxStandardInitExclusiveCPUs {
+					maxStandardInitExclusiveCPUs = cpus
+				}
+			} else {
+				hasSharedStandardInitContainer = true
+			}
+		}
+	}
+
+	// This check ensures that the pool of CPUs available to standard init containers
+	// (the total pod budget minus any sidecars) is large enough to satisfy the
+	// request of the largest standard init container that requires exclusive CPUs.
+	// Since standard init containers run sequentially, we only need to fit the largest one.
+	if maxStandardInitExclusiveCPUs > poolForStandardInits {
+		return fmt.Errorf("pod rejected, largest exclusive init container requests %d cpus, but only %d available", maxStandardInitExclusiveCPUs, poolForStandardInits)
+	}
+
+	// This check ensures that if there are any standard init containers that need a
+	// shared pool, that pool is not empty. An empty pool would occur if the pod
+	// budget is fully consumed by sidecars.
+	if hasSharedStandardInitContainer && poolForStandardInits == 0 {
+		return fmt.Errorf("pod rejected, pod has shared init containers but no cpus available for them")
+	}
+
+	return nil
+}
+
+// This function is the entry point for pod-level resource allocation.
+// It's called once per pod by the Topology Manager's pod-scope admit handler.
+// The logic here allocates a single "bubble" of CPUs for the entire pod
+// and then partitions that bubble among the containers.
+func (p *staticPolicy) AllocatePod(logger logr.Logger, s state.State, pod *v1.Pod, hint topologymanager.TopologyHint) error {
```

Comment:
we need full unit test coverage (= all the flows) for this new function plus e2e test coverage for the main flows.
Full unit test coverage here is a alpha-blocking requirement; e2e test coverage is strongly preferred but we have some wiggle room. Please take into account that if you plan to move quickly to beta *and not* to have a second alpha, then the earler we add the e2e tests (and the more we have till good coverage) the better.

Reply:
It seems that till now you added unit test coverage for `validatePodScopeResources`, which is a great start but not sufficient